src/req_rep.rs

```diff
@@ -25,21 +28,43 @@ pub struct ReqSocket {
 #[async_trait]
 impl BlockingSend for ReqSocket {
     async fn send(&mut self, message: ZmqMessage) -> ZmqResult<()> {
-        let frames = vec![
-            "".into(), // delimiter frame
-            message,
-        ];
-        // TODO this is supposed to be round robin
-        for mut peer in self.backend.peers.iter_mut() {
-            peer.send_queue
-                .send(Message::MultipartMessage(frames))
-                .await?;
-            self.current_request = Some(peer.identity.clone());
-            return Ok(());
+        if self.current_request.is_some() {
+            return Err(ZmqError::Socket(
+                "Unable to send message. Request already in progress",
+            ));
+        }
+        // In normal scenario this will always be only 1 iteration
+        // There can be special case when peer has disconnected and his id is still in RR queue
+        // This happens because SegQueue don't have an api to delete items from queue.
+        // So in such case we'll just pop item and skip it if we don't have a matching peer in peers map
+        loop {
+            let next_peer_id = match self.backend.round_robin.pop() {
+                Ok(peer) => peer,
+                Err(_) => {
+                    return Err(ZmqError::Other(
+                        "Not connected to peers. Unable to send messages",
+                    ))
+                }
+            };
+            match self.backend.peers.get_mut(&next_peer_id) {
+                Some(mut peer) => {
+                    self.backend.round_robin.push(next_peer_id.clone());
+                    let frames = vec![
+                        "".into(), // delimiter frame
+                        message,
+                    ];
+                    peer.send_queue
+                        .send(Message::MultipartMessage(frames))
+                        .await?;
+                    self.backend
+                        .current_request_peer_id
+                        .store(self.backend.peers.hash_usize(&next_peer_id));
```

Comment:
This should be fine, but `hash_usize()` is *probably* not on par with cryptographic hash functions like SHA-1, so I'd expect the probability of collisions to be higher. IIUC, the only reason you are going for using of the hash function in the first place is that you want to use an atomic, but peer identity is a vector of u8's and can't be used as one. Perhaps, an alternative would be to map peer identity to an incrementing integer and use that instead.